integrations/MoEngage/browser.js

```diff
@@ -0,0 +1,199 @@
+          return t;
+        };
+      a.onload = function () {
+        if (n) {
+          i[r] = moe(n);
+        }
+      };
+    })(
+      window,
+      document,
+      "script",
+      "https://cdn.moengage.com/webpush/moe_webSdk.min.latest.js",
+      "Moengage"
+    );
+
+    if (this.region !== "US") {
+      self.moeClient = window.moe({
+        app_id: this.apiId,
+        debug_logs: this.debug ? 1 : 0,
+        cluster: this.region === "EU" ? "eu" : "in",
+      });
+    } else {
+      self.moeClient = window.moe({
+        app_id: this.apiId,
+        debug_logs: this.debug ? 1 : 0,
+      });
+    }
+    this.initialAnonId = window.rudderanalytics.getAnonymousId();
+  }
+
+  isLoaded = () => {
+    logger.debug("in MoEngage isLoaded");
+    return !!window.moeBannerText;
+  };
+
+  isReady = () => {
+    logger.debug("in MoEngage isReady");
+    return !!window.moeBannerText;
+  };
+
+  track(rudderElement) {
+    logger.debug("inside track");
+    // Check if the anonymous id is same as previous session if not a new session will start
+    if (rudderElement.message) {
+      const { anonymousId, event, properties } = rudderElement.message;
+      if (anonymousId) {
+        if (this.initialAnonId !== anonymousId) {
+          this.reset();
+        }
+      }
+      if (event) {
+        if (properties) {
+          this.moeClient.track_event(event, properties);
+        } else {
+          this.moeClient.track_event(event);
+        }
+      } else {
+        logger.error("Event name not present");
+      }
+    } else {
+      logger.error("Payload not correct");
+    }
+  }
+
+  reset() {
+    logger.debug("inside reset");
+    // reset the anonymous id
+    this.initialAnonId = window.rudderanalytics.getAnonymousId();
+    this.moeClient.destroy_session();
+  }
+
+  identify(rudderElement) {
+    const self = this;
+    const { anonymousId, userId } = rudderElement.message;
+    const { traits } = rudderElement.message.context;
+    // check if anonymous id is same or not
+    if (this.initialAnonId !== anonymousId) {
+      this.reset();
+    }
+    // if user is present map
+    if (userId) {
+      this.moeClient.add_unique_user_id(userId);
+    }
+    // custom traits mapping context.traits --> moengage properties
+    const traitsMap = {
+      firstName: "first_name",
+      lastName: "last_name",
+      email: "email",
+      phone: "mobile",
+      name: "user_name",
+      username: "user_name",
+      gender: "gender",
+      birthday: "birthday",
+      id: null,
+    };
+    if (traits) {
+      each(function name(value, key) {
+        // check if name is present
+        if (key === "name") {
+          self.moeClient.add_user_name(value);
+        }
+
+        // if any of the custom properties are present it will be sent
+        // For example : email is present in username then the method will be add_user_name
+
+        if (Object.prototype.hasOwnProperty.call(traitsMap, key)) {
```

Comment:
can the below 3 loops be merged into one?
perform lookup to traitsMap, if present use **add${key_name}** else use **add_user_attribute**

Reply:
but we have to loop through traits too right? to send the remaining attributes?